### PR TITLE
ActivityNotFoundException when calling getBraveActivity

### PR DESCRIPTION
**Pull request by sujitacharya2005**
Resolves https://github.com/brave/brave-browser/issues/27121

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 1]
```diff
@@ -1715,7 +1716,7 @@ static public BraveActivity getBraveActivity() {
             return (BraveActivity)ref;
         }
 
-        return null;
+        throw new ActivityNotFoundException("BraveActivity Not Found");
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1719]: The idea was to mark function `throws ActivityNotFoundException` to enforce handling this case everywhere we call it.

**Review by samartnik — CHANGES_REQUESTED** [review]

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 1 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:1719]: as @samartnik mentioned we need to catch `ActivityNotFoundException` everywhere we use getBraveActivity.

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java [hunk 2]
```diff
@@ -113,12 +114,15 @@ public static Pair<String, String> getBraveStatsStringFormNumberPair(
...
 
     public static void showBraveStats() {
-        if (BraveActivity.getBraveActivity() != null) {
-            BraveStatsBottomSheetDialogFragment braveStatsBottomSheetDialogFragment =
-                    BraveStatsBottomSheetDialogFragment.newInstance();
-            braveStatsBottomSheetDialogFragment.show(
-                    BraveActivity.getBraveActivity().getSupportFragmentManager(),
-                    STATS_FRAGMENT_TAG);
+        try {
+            if (BraveActivity.getBraveActivity() != null) {
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:118]: the checks on null could be deleted everywhere as we don't return null anymore

File: android/java/org/chromium/chrome/browser/BraveAdFreeCalloutDialogFragment.java [hunk 3]
```diff
@@ -58,10 +59,12 @@ public void onViewCreated(View view, Bundle savedInstanceState) {
 
         Button btnVideos = view.findViewById(R.id.btn_videos);
         btnVideos.setOnClickListener(button -> {
-            BraveActivity braveActivity = BraveActivity.getBraveActivity();
-            if (braveActivity != null) {
+            try {
+                BraveActivity braveActivity = BraveActivity.getBraveActivity();
                 braveActivity.focusSearchBox();
+            } catch (ActivityNotFoundException e) {
             }
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/BraveAdFreeCalloutDialogFragment.java:66]: Would be nice to have a log in each such cases to make debugging easier.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/NetworkSelectorActivity.java [hunk 4]
```diff
@@ -71,10 +71,12 @@ public void finishNativeInitialization() {
     }
 
     private void initState() {
-        BraveActivity activity = BraveActivity.getBraveActivity();
-        if (activity != null) {
+        try {
+            BraveActivity activity = BraveActivity.getBraveActivity();
             mWalletModel = activity.getWalletModel();
+        } catch (Exception e) {
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/NetworkSelectorActivity.java:77]: Should be ActivityNotFoundException

File: android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java [hunk 5]
```diff
@@ -529,13 +530,17 @@ public void onPageLoadFinished(final Tab tab, GURL url) {
...
+                    if (BraveActivity.getBraveActivity() != null) {
+                        RewardsYouAreNotEarningDialog rewardsYouAreNotEarningDialog =
+                                RewardsYouAreNotEarningDialog.newInstance();
+                        rewardsYouAreNotEarningDialog.setCancelable(false);
+                        rewardsYouAreNotEarningDialog.show(
+                                BraveActivity.getBraveActivity().getSupportFragmentManager(),
+                                RewardsYouAreNotEarningDialog.RewardsYouAreNotEarningDialogTAG);
+                    }
+
+                } catch (Exception e) {
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java:543]: Should be ActivityNotFoundException

File: browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java [hunk 6]
```diff
@@ -116,10 +116,12 @@ public void onClick(DialogInterface dialog, int which) {
...
-
-                // Enable ads
-                if (BraveActivity.getBraveActivity() != null) {
-                    BraveActivity.getBraveActivity().openRewardsPanel();
+                try {
+                    // Enable ads
+                    if (BraveActivity.getBraveActivity() != null) {
+                        BraveActivity.getBraveActivity().openRewardsPanel();
+                    }
+                } catch (Exception e) { // sujit
```
Comment (samartnik) [inline browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java:124]: Should be ActivityNotFoundException. Also not sure what this comment implies.

File: browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java [hunk 7]
```diff
@@ -142,9 +144,12 @@ public static void showExistingUserDialog(Context context) {
...
                 neverShowOnboardingDialogAgain();
-                BraveActivity braveActivity = BraveActivity.getBraveActivity();
-                if (braveActivity != null) {
-                    braveActivity.openRewardsPanel();
+                try {
+                    BraveActivity braveActivity = BraveActivity.getBraveActivity();
+                    if (braveActivity != null) {
+                        braveActivity.openRewardsPanel();
+                    }
+                } catch (Exception e) { // sujit
```
Comment (samartnik) [inline browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java:152]: Should be ActivityNotFoundException. Also not sure what this comment implies.

**Review by samartnik — CHANGES_REQUESTED** [review]

File: android/java/org/chromium/chrome/browser/util/TabUtils.java [hunk 8]
```diff
@@ -175,50 +187,63 @@ private static void openNewTab(BraveActivity braveActivity, boolean isIncognito)
...
 
     public static void bringChromeTabbedActivityToTheTop(Activity activity) {
-        Intent braveActivityIntent = new Intent(activity, ChromeTabbedActivity.class);
-        braveActivityIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-        activity.startActivity(braveActivityIntent);
+        try {
+            Intent braveActivityIntent = new Intent(activity, ChromeTabbedActivity.class);
+            braveActivityIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+            activity.startActivity(braveActivityIntent);
+        } catch (ActivityNotFoundException e) {
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/util/TabUtils.java:245]: why would we add try catch in that place?

**Comment by SergeyZhukovsky** [comment]
I would also add logs in each catch as @samartnik suggested and asserts probably so it triggers in debug configuration and we can take a look what's wrong.

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 1 above)
Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:1673]: updated

File: android/java/org/chromium/chrome/browser/BraveAdFreeCalloutDialogFragment.java (hunk 3 above)
Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/BraveAdFreeCalloutDialogFragment.java:70]: updated

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/NetworkSelectorActivity.java (hunk 4 above)
Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/NetworkSelectorActivity.java]: updated

File: android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java (hunk 5 above)
Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java]: updated

File: browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java (hunk 6 above)
Comment (sujitacharya2005) [reply browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java]: updated

File: browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java (hunk 7 above)
Comment (sujitacharya2005) [reply browser/brave_ads/android/java/org/chromium/chrome/browser/dialogs/BraveAdsSignupDialog.java]: updated

File: android/java/org/chromium/chrome/browser/util/TabUtils.java (hunk 8 above)
Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/util/TabUtils.java]: updated

**Comment by sujitacharya2005** [comment]
> I would also add logs in each catch as @samartnik suggested and asserts probably so it triggers in debug configuration and we can take a look what's wrong.

@SergeyZhukovsky Updated

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Review by samartnik — APPROVED** [review]
lgtm

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/NftDetailActivity.java [hunk 9]
```diff
@@ -44,6 +46,8 @@
 import java.util.Locale;
 
 public class NftDetailActivity extends BraveWalletBaseActivity {
+    public static final String TAG = "NftDetailActivity";
```
Comment (aseren) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/NftDetailActivity.java:49]: maybe `private static final`?

File: android/java/org/chromium/chrome/browser/util/TabUtils.java [hunk 10]
```diff
@@ -175,44 +193,58 @@ private static void openNewTab(BraveActivity braveActivity, boolean isIncognito)
...
-        View toolbarView = braveActivity.findViewById(R.id.toolbar);
-        if (toolbarView == null) {
-            return;
-        }
-        FrameLayout rewardsLayout = toolbarView.findViewById(R.id.brave_rewards_button_layout);
-        if (rewardsLayout == null) {
-            return;
+        try {
+            BraveActivity braveActivity = BraveActivity.getBraveActivity();
+            if (braveActivity == null || braveActivity.getToolbarManager() == null) {
```
Comment (aseren) [inline android/java/org/chromium/chrome/browser/util/TabUtils.java:233]: Is `braveActivity == null` check required here?

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 11]
```diff
@@ -1662,14 +1663,14 @@ static public ChromeTabbedActivity getChromeTabbedActivity() {
         return null;
     }
 
-    static public BraveActivity getBraveActivity() {
+    static public BraveActivity getBraveActivity() throws ActivityNotFoundException {
```
Comment (aseren) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1666]: should we add @NotNull annotation to this method?

Comment (sujitacharya2005) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:1667]: updated

**Review by aseren — APPROVED** [review]

**Pull request merged**